Signup

gonziesc opened the pull request:
![travis](https://travis-ci.org/wolox-training/ge-express-js.svg?branch=signin)
## Summary
- User can request (POST) to “/users”
- Validated email with Wolox domain
- Validated email uniqueness
- Validated password length
- Auto-incremental unique id
- Tests

## Screenshots
![screenshot from 2017-04-17 16-16-11](https://cloud.githubusercontent.com/assets/5241126/25101118/5a365124-2389-11e7-8f30-d1859143df67.png)
![screenshot from 2017-04-17 16-16-26](https://cloud.githubusercontent.com/assets/5241126/25101117/5a35ee1e-2389-11e7-9af4-015f1072f3bc.png)
![screenshot from 2017-04-17 16-16-37](https://cloud.githubusercontent.com/assets/5241126/25101116/5a35d906-2389-11e7-828d-219cd7ccbbe6.png)
![screenshot from 2017-04-17 16-16-47](https://cloud.githubusercontent.com/assets/5241126/25101120/5a3934e8-2389-11e7-9db1-ca42e1112922.png)
![screenshot from 2017-04-17 16-16-58](https://cloud.githubusercontent.com/assets/5241126/25101115/5a3555a8-2389-11e7-99d3-ec84f476e1a4.png)


## Trello card
https://trello.com/c/nSdgOh8K/2-sign-up

mishuagopian commented on app.js:
```diff
@@ -5,7 +5,8 @@ const express = require('express'),
   orm = require('./app/orm'),
-  errors = require('./app/middlewares/errors');
+  errors = require('./app/middlewares/errors'),
+  validator = require('email-validator');
```
Remove this line, it's unused.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+'use strict';
```
Delete.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+'use strict';
+
+const userService = require('../services/users'),
+  validator = require('email-validator'),
```
I'd validate emails with a regex. Let's avoid packages when we don't really need them.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+  } else {
+    errorsReturn += 'The mail field can not be empty <br/>';
+  }
+  return errorsReturn;
```
As we are developing a json API, we are returning a json object. Because of this, we can return an array of strings with the errors instead of a html-formatted string.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+
+exports.create = (req, res, next) => {
+  const saltRounds = 10;
+  const passw = req.param('password');
```
I think `req.param('something')` is outdated. Replace it with `req.body.something`.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+};
+
+exports.create = (req, res, next) => {
+  const saltRounds = 10;
```
unused

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+  validator = require('email-validator'),
+  errors = require('../errors');
+
+const validateUser = function (name, lastName, mail) {
```
I'd move this method to a service like `/services/users`

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+      next(errors.defaultError(err));
+    });
+  } else {
+    res.status(501);
```
5** are for internal server errors, use 400 for validation errors. Remember sending the error object!

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,57 @@
+  const surname = req.param('lastName');
+  const mail = req.param('email');
+  let errorsSend = validateUser(name, surname, mail);
+  if (passw.length <= 8) {
```
I'd validate the password inside `validateUser` too.

mishuagopian commented on app/models/models.js:
```diff
@@ -1,5 +1,4 @@
-// const model = require('./model');
-
+const user = require('./user');
```
enter

mishuagopian commented on app/models/scripts/dataCreation.js:
```diff
@@ -1,6 +1,24 @@
-
+  const data = [];
+  const users = [];
+  users.push(db.models.user.create({
```
Push users to `data` array as is the promise you are returning.

mishuagopian commented on config/testing.js:
```diff
@@ -2,7 +2,7 @@ exports.config = {
   common: {
     database: {
-      url: process.env.NODE_API_DB_URL,
+      url: process.env.NODE_API_DB_URL_TEST,
```
Would make a PR to the [express-js-bootstrap](http://www.github.com/wolox/express-js-bootstrap) with this change ?

mishuagopian commented on package.json:
```diff
@@ -18,8 +18,10 @@
     "start": "./node_modules/.bin/devtool app.js --watch --browser-field",
-    "test": "NODE_ENV=testing ./node_modules/mocha/bin/mocha --compilers js:babel-core/register test/app.js && npm run nsp"
+    "test": "NODE_ENV=testing ./node_modules/mocha/bin/mocha --compilers js:babel-core/register test/app.js && npm run nsp",
+    "users": " mocha"
```
Why this?

mishuagopian commented on test/users.js:
```diff
@@ -0,0 +1,71 @@
+  server = require('./../app'),
+  should = chai.should();
+
+describe('/users POST', () => {
```
I like these test cases. I'd also verify error messages in every test.

gonziesc commented:
![travis](https://travis-ci.org/wolox-training/ge-express-js.svg?branch=signup)
## Summary 
- validated error messages as strings
- deleted unused lines
- changed promise push to users on data creation
- verified error messages in tests

## Trello Card
https://trello.com/c/nSdgOh8K/2-sign-up

mishuagopian commented on app/services/users.js:
```diff
@@ -1,8 +1,34 @@
+    }
+    if (domain !== 'wolox.com.ar') {
+      errorsReturn.push('The domain has to be Wolox');
+    }
```
This looks cleaner:
```node
/^[^@\s]+@wolox.com.ar$/i.test(email)
```

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,31 @@
+  const passw = req.body.password;
+  const name = req.body.firstName;
+  const surname = req.body.lastName;
+  const mail = req.body.email;
```
Use email, not mail.

mishuagopian replied:
Maybe you can user firstName, lastName, email and password so that we use es6 syntax to create the object.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,31 @@
+  errors = require('../errors');
+
+exports.create = (req, res, next) => {
+  const passw = req.body.password;
```
Use the whole word for variables: `password` here.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,31 @@
+  const name = req.body.firstName;
+  const surname = req.body.lastName;
+  const mail = req.body.email;
+  const errorsSend = userService.validateUser(name, surname, mail, passw);
```
Just `errors` and not `errorsSend`.

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,31 @@
+  const surname = req.body.lastName;
+  const mail = req.body.email;
+  const errorsSend = userService.validateUser(name, surname, mail, passw);
+  if (passw.length <= 8) {
```
Why not validating password inside the service?

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,31 @@
+  if (passw.length <= 8) {
+    errorsSend.push('The password has to have 8 or more characters');
+  }
+  if (errorsSend.length === 0) {
```
This will be much cleaner:
```node
if (errors) {
  res.status(400); // this is a 400 and not a 401 error
  res.end();
} else {
  userService.create({ firstName, lastName, password, email }).then(() => {
    res.status(200);
    res.end();
  }).catch(next);
}
```

mishuagopian commented on app/models/user.js:
```diff
@@ -0,0 +1,33 @@
+      allowNull: false,
+      unique: true
+    },
+    password: {
```
min length here?

mishuagopian commented on app/services/users.js:
```diff
@@ -0,0 +1,33 @@
+  });
+};
+
+exports.validateUser = function (name, lastName, mail) {
```
What about the password here?

mishuagopian replied:
use `(name, lastName, email) => {` instead of `function (name, lastName, email) {`

mishuagopian commented on app/services/users.js:
```diff
@@ -0,0 +1,33 @@
+};
+
+exports.validateUser = function (name, lastName, mail) {
+  const errorsReturn = [];
```
Just `errors` for this name

mishuagopian commented on test/users.js:
```diff
@@ -0,0 +1,82 @@
+    .catch((err) => {
+      err.should.have.status(401);
+      const errorsRecieved = err.response.text;
+      errorsRecieved.should.be.equal('["The mail field can not be empty"]');
```
Check for values inside the array, not to the string value of the returned object.

gonziesc commented:
![travis](https://travis-ci.org/wolox-training/ge-express-js.svg?branch=signup)
## Summary 
- checked values inside arrays in tests
- min length added to model
- created user object
- used regex to validate mail
- changed error to 400

## Trello Card
https://trello.com/c/nSdgOh8K/2-sign-up

mishuagopian commented on app/services/sessionManager.js:
```diff
@@ -0,0 +1,22 @@
+  const expiringDate = new Date();
+  expiringDate.setDate(expiringDate.getDate() + 7);
+  const claims = { sub: email,
+    iss: '/users/sessions',
```
remove

mishuagopian commented on app/services/sessionManager.js:
```diff
@@ -0,0 +1,22 @@
+    exp: expiringDate,
+    permissions: 'access'
+  };
+  const token = exports.encode(claims);
```
`return exports.encode(claims);`

mishuagopian commented on app/controllers/users.js:
```diff
@@ -0,0 +1,22 @@
+    userService.create(user).then((u) => {
+      res.status(200);
+      res.end();
+    }).catch((err) => {
```
```node
}).catch(next);
```

mishuagopian commented on app/services/sessionManager.js:
```diff
@@ -0,0 +1,22 @@
+  const claims = { sub: email,
+    iss: '/users/sessions',
+    exp: expiringDate,
+    permissions: 'access'
```
remvoe

mishuagopian commented on app/services/sessionManager.js:
```diff
@@ -0,0 +1,22 @@
+exports.generateAccessToken = (email) => {
+  const expiringDate = new Date();
+  expiringDate.setDate(expiringDate.getDate() + 7);
+  const claims = { sub: email,
```
Why sub and not just email ?

gonziesc commented:
![travis](https://travis-ci.org/wolox-training/ge-express-js.svg?branch=signup)
## Summary 
- removed unused code
- changed function to lambda

## Trello Card
https://trello.com/c/nSdgOh8K/2-sign-up

picollomartin commented on app/routes.js:
```diff
@@ -1,8 +1,10 @@
   app.get('/health', healthCheck);
   app.get('/albums', [], albumController.getAlbums);
   app.get('/albums/:id/photos', [], albumController.getAlbumPhotos);
+  app.post('/users', [], userController.signUp);
```
Remove empty array of middlewares, `app.post('/users', userController.signUp);`, is not necessary

picollomartin commented on app/controllers/user.js:
```diff
@@ -0,0 +1,20 @@
+    return res.status(400).send(validationErrors);
+  }
+  return encrypt(req.body.password)
+    .then(encryptedPassword => createUser({ ...req.body, password: encryptedPassword }))
```
Be careful with spread body and send directly to database, if i send extra fields could be persisted whatever i want

picollomartin commented on app/controllers/user.js:
```diff
@@ -0,0 +1,20 @@
+  return encrypt(req.body.password)
+    .then(encryptedPassword => createUser({ ...req.body, password: encryptedPassword }))
+    .then(([user, created]) =>
+      res.status(created ? 201 : 409).send(created ? user : ['Email already registered'])
```
Use middleware of errors to handle response errors, this map a error internal code to a status code and a message

picollomartin commented on app/controllers/user.js:
```diff
@@ -0,0 +1,20 @@
+    )
+    .catch(err => {
+      logger.error(`Error creating user: ${err}`);
+      res.status(500).send(err);
```
Same here, this could be done like `next(err)`

picollomartin commented on app/utils/user.js:
```diff
@@ -0,0 +1,25 @@
+  } else if (!data.email.match(/\S+@\S+\.\S+/)) {
+    errors.push('Email format invalid');
+  }
+  return errors;
```
This is ok for training scope (in projects we use something like [this](https://github.com/express-validator/express-validator)), avoid some logic repetition (like check precense of a key in a object, or check a regex)

picollomartin commented on app/controllers/user.js:
```diff
@@ -0,0 +1,20 @@
+exports.signUp = (req, res) => {
+  const validationErrors = validateUserData(req.body);
+  if (validationErrors.length) {
+    return res.status(400).send(validationErrors);
```
Validation error should be handled with middleware of errors too

picollomartin commented on app/utils/encrypt.js:
```diff
@@ -0,0 +1,13 @@
+const bcrypt = require('bcryptjs');
+
+const saltRounds = 10;
```
This could be configurable with a default value `10`

picollomartin reviewed (changes requested):
Nice code 🎉 , some little fixes and we are done.

picollomartin commented on config/index.js:
```diff
@@ -36,7 +36,8 @@ const config = {
       parameterLimit: process.env.API_PARAMETER_LIMIT,
-      port: process.env.PORT
+      port: process.env.PORT,
+      saltRounds: parseInt(process.env.SALT_ROUNDS)
```
Move this outside API scope

picollomartin commented on app/utils/encrypt.js:
```diff
@@ -0,0 +1,17 @@
+const bcrypt = require('bcryptjs'),
+  {
+    common: {
+      api: { saltRounds = DEFAULT_SALT_ROUNDS }
```
It's okay but is a little tricky to define a default value, i prefer something like 
``` 
const config = require('../../config');
const saltRounds = config.salt || DEFAULT_VALUE;
```

picollomartin reviewed (approved):
Nice 🚀
